### ISSUE TITLE
Feature: locationUpdateInterval for iOS

### ISSUE THREAD

**Issue by filippopossenti** [transistorsoft/cordova-background-geolocation-lt#643]
Extend compatibility of "locationUpdateInterval" to iOS devices, to better control granularity of data sampling in certain business scenarios.

## Environment
* Plugin version: master trunk
* Platform: iOS
* OS version: 11.3
* Device manufacturer / model: Apple iPhone 6
* Cordova version (`cordova -v`): 7.1
* Cordova platform version (`cordova platform ls`): 4.5.3
* Plugin config:
{
				desiredAccuracy: tracking_accuracy,
				distanceFilter: BackgroundGeolocation.DESIRED_ACCURACY_LOW,
				stopOnStationary: false,
				stopOnTerminate: false,
				heartbeatInterval: 60,
				method: "POST",
				httpTimeout: 30000,
				autoSync: true,
				batchSync: true,
				maxBatchSize: 10,
				maxDaysToPersist: 5,
				// Android
				locationUpdateInterval: tracking_interval * 1000,
				fastestLocationUpdateInterval: 10000,
				// foregroundService: true,	// don't do this or you'll have a black screen if the user terminates the app by swiping it away. Worked around through a dev option, but still a problem
				stopOnStillActivity: false,
				allowIdenticalLocations: true,	// even if the location is identical, we still want it to ensure that sampling "frequency" is maintained
				// iOS
				stationaryRadius: BackgroundGeolocation.DESIRED_ACCURACY_LOW,
				useSignificantChangesOnly: false,
				preventSuspend: true
			}

## Expected Behavior
iOS should observe a defined interval between captures (well, the captures that will end up on the internal DB and then sent on the server), as defined by the currently Android-only property "locationUpdateInterval".

## Actual Behavior
Captures are taken at very narrow intervals (few seconds), flooding the database and the target system with data at a resolution that is not desired.
Workarounds like collecting and sending the data manually through JavaScript and Ajax calls have already been evaluated and are deemed not practical.

## Steps to Reproduce
1. just use the plugin on iOS and set "locationUpdateInterval". It won't have any effect.

## Context
Our app is meant for business use. Within its options, the user can configure how often GPS tracking data is to be sent. Such data is used to determine the approximate location of the user in near-realtime as this information is then used over the course of the day to schedule tasks from the office based on the current position of an operative.
As a result, we need updates to span in the range 1-30 minutes. Excessive data collection is undesirable as it translates in too much data being sent to our systems as well as too much data consumed in the user's data plans.

**Comment by christocracy** [comment]
iOS API provides no such mechanism.  The CLLocationManager api is *strictly* distance-based.  Time-based tracking is a *bad* idea.

If you find yourself going down this path, you need to re-think your plan.

**Comment by filippopossenti** [comment]
And yet, the Android version does support tweaking the frequency of data collection through "locationUpdateInterval" so for consistency it would be nice to have something similar on iOS as well. Mind that this doesn't mean it should keep tracking you all the time regardless of significant changes, it just means that when the GPS kicks in and starts collecting data, instead of blindly put it in the database and send it to the server it would be helpful to provide a parameter that allows to control whether the new data sample should be collected (based on how old the previous sample is) and therefore sent.

While I do agree that distance-based is better, I also have to take exception and maintain that time-based (which, by the way, it's not: it's just "sample throttling") is a perfectly valid options in specific scenarios, for a number of reasons that are not necessarily related to development choices. Examples: a previous app works with time-based sampling, meaning that a new app forces to retrain users... which has a cost. No longer being able to rely on a configurable sampling frequency forces to redesign also some server-side features... which has a cost.
So, you see, distance-based might be better from developer's perspective and it's certainly something to push towards, it's not necessarily feasible in the scope of every project.

If you prefer to see it in a different way, see it like this: it's all good as it is, but it would be useful being able to say "don't collect a new sample if the previous one is not X milliseconds old".

**Comment by christocracy** [comment]
You’re free to use the plugin’s #watchPosition method / preventSuspend as desired, and at your own risk.

There will **never** be a `locationUpdateInterval` option for iOS unless iOS provides it on their api in a future os version (highly doubtful)

**Comment by filippopossenti** [comment]
Yes, watchPosition will do. I didn't think about it.
At my own risk of course... and I can see how bad it will be on the battery.

**Comment by christocracy** [comment]
I would be `#watchPosition` / `#stopWatchPosition` in `motionchange` event when device is moving / stationary.  Otherwise you're basically turning on GPS 24/7.   The battery will die within hours.

**Comment by filippopossenti** [comment]
Yes, I was thinking about that too.

**Comment by christocracy** [comment]
Have you read the *Philosophy of Operation* in Wiki?

**Comment by filippopossenti** [comment]
Yes, superficially when I first saw this plugin time ago and more thoroughly now.
It won't be an easy one, especially on iOS... but in the end I'm confident it will work.

The ticket can be closed if you wish. Still, consider adding the feature I proposed. Not to be interpreted as "locationUpdateInterval" as that doesn't exist but as a "locationPersistInterval".

**Comment by filippopossenti** [comment]
Sorry, just to further specify my idea:
it's more like emulating "fastestLocationUpdateInterval" on iOS by saving the location on the DB only once a new location is past the previous+interval point in time.

**Comment by christocracy** [comment]
What is your concern with this?

(btw, this config option makes no sense:  `stationaryRadius: BackgroundGeolocation.DESIRED_ACCURACY_LOW,`)

**Comment by filippopossenti** [comment]
As explained before, the point of the feature is to allow a uniform behaviour between iOS and Android when it comes to scenarios where a constant sampling frequency is required.
The APIs you suggested will probably do, but of course having that feature provided directly by the plugin would lift the developer from a lot of testing in such scenarios, especially considering that GPS testing often requires to "go around and have a walk", which implies a lot of dead time. On our end, in this particular case, a choice was made not to go through the APIs suggestion, limit the user options and rely on server-side code to avoid flooding the database (as for the users' data-plan we'll collect telemetry to assess whether it has the potential to become a problem or not).

Regarding the configuration: the value of DESIRED_ACCURACY_LOW is currently 100. That setting effectively states that the distanceFilter is 100m. While I don't think is a particularly good idea to use that constant, I also believe it's value is unlikely to change so I considered the practice acceptable as it helps understanding that 100 is not arbitrary but connected to the fact that a low accuracy is acceptable.

Still, having gone through 3 plugins now, I admit the configuration is not particularly clean and that particular value comes from a precedent plugin. The updated configuration is this:

            var options = {
                // http and persistance parameters
                method: "POST",
                httpTimeout: 30000,
                autoSync: true, // will try to send location data immediately
                batchSync: true,
                maxBatchSize: 10,
                maxDaysToPersist: 5,
                // location parameters (shared)
                desiredAccuracy: tracking_accuracy,
                stopOnStationary: false,
                debug: false, //debug: Platform.App.isDebug(),
                stopOnTerminate: false, // we don't stop unless the user logs out. The reason being that the app could be terminated once in background or the use might swipe it away and we don't want this to affect background tracking
                // Android
                fastestLocationUpdateInterval: tracking_interval * 1000,
                // foregroundService: true, // don't do this or you'll have a black screen if the user terminates the app by swiping it away
                // iOS
                useSignificantChangesOnly: low_energy   // apparently, even if tracking_accuracy is low, on iOS it uses GPS. This will prevent it when on low accuracy
            };

**Comment by filippopossenti** [comment]
"briefly", constant sampling frequency is required because this is not a new app and needs to maintain behaviours our customers are already used to and happy with and our servers are built around.

If you were interested only in the "brief" part stop here.



The purpose of our app is to allow operatives working for our customers to fulfill jobs composed of several steps. Some companies, in order to schedule jobs effectively, need to know where the user is. Sometimes they need a high degree of precision, often they don't. Sometimes they need to know when the user was at a specific point in time, sometimes they don't.
The app is not "new". There's already a solid Android version, first released years ago and as of today constantly maintained and improved. The first and most important objective of the cordova version of the app is to extend compatibility to iOS preserving as much as possible the behaviours that were tuned and tweaked and considered optimal for our specific needs over the course of the years (hence why I "feel" that such feature is required: it's been there forever and it worked well for the purposes it's been used for).


There might be a reason Apple didn't make this easy but there's also a reason why Google instead made it so easy (as the "fastestLocationUpdateInterval" variable clearly points out).

You already stated before that we should re-think our strategy. And I already explained that while I agree "in principle", sometimes you have to go down another path and a strategy is not necessarily "black or white" but might have various shades of gray and may require intermediate steps.

And I disagree with your last statement: you can absolutely develop mobile apps borrowing "specific" web-app paradigms, especially when the app is not meant to work alone but in conjunction with an actual web-app used by users whose job is to stay behind a PC screen and coordinate the work. Besides, we are talking about a framework (cordova) built around a web browser...


Provided that our testing goes well, we plan to buy a license to the plugin. This specific feature is not critical to the purchase but of course having it at some point would help considerably. Whether to implement it or not, of course, is up to the developer and that's you.

**Comment by christocracy** [comment]
I’ve been working on this plugin nearly 5 years.  There are thousands of users.  Your use-case is not unique.

I’m guessing your app is not using “silent” server push requests.

Are you familiar with the plugin’s background-fetch dependency.

**Comment by filippopossenti** [comment]
The guess is correct. Push notifications are a feature we will introduce in the future, although for purposes different than location tracking.

I am familiar with the purposes of the background-fetch plugin, which was and is still considered to address other issues (that is, background synchronisation in the absence of push notifications, exactly).

Unfortunately I've exhausted the alloted to the location features, so I won't be applying any further change to the code driving that particular feature unless the tests show some serious issue.

**Comment by christocracy** [comment]
A “silent” push request is a request from server to a particular device or *set* of devices with only a JSON payload.  The user is completely unaware of the notification; there is no message.

Your JavaScript can receive and decide this JSON and choose to do anything it wishes on the client device.

An iOS push notification is capable of rebooting a terminated device in the background.

Eg:
```
{
  “command”: “getCurrentPosition”
}
```
Your app code would decode this JSON and choose to interact with the `BackgroundGeolocation` api.

A web application user might cause this silent push to be sent to a device from a [Locate] button on the UI.
